### PR TITLE
Notebook launcher

### PR DESCRIPTION
**Pull request by sgugger**
This PR adds a notebook launcher function, to easily launch training in a Colab with TPUs.
having it properly launch distributed training in a Notebook on a machine with several GPUs seems a bit less straightforward, so support will be added later if there interest from the community.

Notebook examples will follow!

### CODE REVIEW

File: docs/source/launcher.rst [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+    Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+    an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+    specific language governing permissions and limitations under the License.
+
+
+Notebook Launcher
+=======================================================================================================================
+
+Launch your training function from a notebook. Currently supports launching a training with TPUs on Colab, or training
```
Comment (philschmid) [inline docs/source/launcher.rst:17]:
```suggestion
Launch your training function inside a notebook. Currently supports launching a training with TPUs on [Google Colab](https://colab.research.google.com/), or training
```
```suggestion
Launch your training function from a notebook. Currently supports launching a training with TPUs on Colab, or training
```

File: src/accelerate/notebook_launcher.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    .. warning::
+
+        Multiple GPUs is not yet supported.
+    """
+    launcher = PrepareForLaunch(function)
+
+    # Are we in a google colab?
+    if "IPython" in sys.modules:
+        in_colab = "google.colab" in str(sys.modules["IPython"].get_ipython())
+    else:
```
Comment (philschmid) [inline src/accelerate/notebook_launcher.py:53]: We could add a check for Kaggle here too. I guess users expect the same behavior there. I checked a Kaggle TPU notebook, sadly `sys.modules["IPython"].get_ipython()` is not hinting that the notebook is run on Kaggle. But we could add a check like `if any(key.startswith('KAGGLE') for key in os.environ.keys())`. Additionally here is a dump of the env vars on kaggle

```python
environ{'LC_ALL': 'C.UTF-8',
        'LD_LIBRARY_PATH': '/opt/conda/lib',
        'KAGGLE_DATA_PROXY_TOKEN': 'eyJhbGciOiJBMTI4S1ciLCJlbmMiOiJBMTI4Q0JDLUhTMjU2In0.B5s-CblnZQjaC_KXAizqBneeEiZmD-pS7Asgb2h-jS6uEkmkIIHLYg.ddiwW2QpGlaxa5Q584WQOg.E8mrT0sr5vZsIb5v4zdJf6uvrzk2wuZCaM6g-xLBTkZo_Nx-lLLqcZqUrZixTOznQqeidkc5tJibiyYBeBGcRnwSuFXx0w4V0EJPfKVqp_B5TMbOROvjnLIpX3eoINPjQd7I38nmCeyrlGUnC-aTHLFqDAxaaYmcESykEixt_Hlc7Q_DfiD_uPkEzJs8buJBvcYGFq8FWbJB7n89rkDY-P9xvZnSsFjbiQYtAbfwcKGdxGKYKAoVdqRhsRHjVjj6HDVQU38Ldnuhzjg2kdt8Xhu__YxSXq0Gh-RFLZxP90M.ZPazCS7UZW9RyI3-jniqPA',
        'LANG': 'C.UTF-8',
        'HOSTNAME': '6497510c329c',
        'TESSERACT_PATH': '/usr/bin/tesseract',
        'XRT_TPU_CONFIG': 'tpu_worker;0;10.0.0.2:8470',
        'KAGGLE_KERNEL_INTEGRATIONS': '',
        'PROJ_LIB': '/opt/conda/share/proj',
        'KAGGLE_KERNEL_RUN_TYPE': 'Interactive',
        'MKL_THREADING_LAYER': 'GNU',
        'NO_GCE_CHECK': 'True',
        'PWD': '/kaggle/working',
        'HOME': '/root',
        'DL_ANACONDA_HOME': '/opt/conda',
        'ANACONDA_PYTHON_VERSION': '3.7',
        'KAGGLE_URL_BASE': 'https://www.kaggle.com',
        'KAGGLE_USER_SECRETS_TOKEN': 'eyJhbGciOiJkaXIiLCJlbmMiOiJBMTI4Q0JDLUhTMjU2In0..kvSYrK_dnCLGYSW8GDnQdg.xlHt52wzWG2SlyvAN6zUq9M2S3m6maqeK_nGH6zCEL-VDZE-CYAC19Ecal8oy7zJQCFYcTDdFGi-Qq6c3tdXinuUy-kG1UQ8MnTSDJJT1AvIWAeQf2MNacS8TMWa7IYG.DLWrZFE6rlAGB7rKowpPKQ',
        'KAGGLE_DOCKER_IMAGE': 'gcr.io/kaggle-images/python:tpu',
        'CONTAINER_NAME': 'base-cpu',
        'KAGGLE_DATA_PROXY_PROJECT': 'kaggle-161607',
        'SHELL': '/bin/bash',
        'MPLBACKEND': 'agg',
        'SHLVL': '1',
        'PYTHONPATH': '/kaggle/lib/kagglegym:/kaggle/lib',
        'CONTAINER_URL': 'gcr.io/deeplearning-platform-release/base-cpu:nightly-2021-03-28',
        'KAGGLE_CONTAINER_NAME': 'kaggle_gq54lOn7oztYNC879IY4P2ItqdI6zTyAKVlolN6KbYE-60764729-webtier',
        'PYTHONUSERBASE': '/root/.local',
        'PATH': '/opt/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin',
        'KAGGLE_DATA_PROXY_URL': 'https://dp.kaggle.net',
        'TPU_NAME': 'grpc://10.0.0.2:8470',
        '_': '/opt/conda/bin/jupyter',
        'GIT_PYTHON_REFRESH': 'quiet',
        'JPY_PARENT_PID': '9',
        'TERM': 'xterm-color',
        'CLICOLOR': '1',
        'PAGER': 'cat',
        'GIT_PAGER': 'cat'}
``` 

Additionally, if we add Kaggle we should rename `in_colab` to maybe `in_colab_or_kaggle`.

**Review by philschmid — COMMENTED** [review]
Very nice! While doing this we could also add support for Kaggle and maybe we should create a colab notebook with an example? It might be a bit hard to understand for not that experienced user what you mean with 
```bash
If you are training on Colab with TPUs, your :obj:`Accelerator` object should only be defined inside the training function. This is because the initialization should be done inside the launcher only.
 ```

**Comment by sgugger** [comment]
Yes, notebook examples will follow! Working on it this afternoon and hoping to have most NLP tasks by the end of the week!

**Pull request merged**